results/cmd/watcher/main_test.go

```diff
@@ -0,0 +1,121 @@
+package main
+
+import (
+	"context"
+	"strings"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	pipelineclient "github.com/tektoncd/pipeline/pkg/client/injection/client"
+	taskruninformer "github.com/tektoncd/pipeline/pkg/client/injection/informers/pipeline/v1beta1/taskrun"
+	ttesting "github.com/tektoncd/pipeline/pkg/reconciler/testing"
+	test "github.com/tektoncd/pipeline/test"
+
+	_ "github.com/mattn/go-sqlite3"
+	"github.com/tektoncd/experimental/results/pkg/server"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+	"knative.dev/pkg/configmap"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/logging"
+)
+
+var (
+	srv *server.Server
+)
+
+func getRunName(tr *v1beta1.TaskRun) string {
+	return strings.Join([]string{tr.Namespace, tr.Name}, "/")
+}
+
+// getFakeClients create a fake client to send test data to reconciler
+func getFakeClients(t *testing.T, tr []*v1beta1.TaskRun) (test.Assets, context.CancelFunc) {
+	t.Helper()
+	ctx, _ := ttesting.SetupFakeContext(t)
+	ctx, cancel := context.WithCancel(ctx)
+	d := test.Data{
+		TaskRuns: tr,
+	}
+	clients, _ := test.SeedTestData(t, ctx, d)
+	cmw := configmap.NewInformedWatcher(clients.Kube, "")
+	return test.Assets{
+		Controller: newController(ctx, cmw, t),
+		Clients:    clients,
+	}, cancel
+}
+
+// newController creates a Controller with provided context and configmap
+func newController(ctx context.Context, cmw configmap.Watcher, t *testing.T) *controller.Impl {
+	logger := logging.FromContext(ctx)
+	taskRunInformer := taskruninformer.Get(ctx)
+	pipelineclientset := pipelineclient.Get(ctx)
+	c := &reconciler{
+		logger:            logger,
+		taskRunLister:     taskRunInformer.Lister(),
+		fakeserver:        srv,
+		pipelineclientset: pipelineclientset,
+	}
+	impl := controller.NewImpl(c, c.logger, pipeline.PipelineRunControllerName)
+	taskRunInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    impl.Enqueue,
+		UpdateFunc: controller.PassNew(impl.Enqueue),
+	})
+	return impl
+}
+
+// getNewTaskRun sends TaskRun data to reconciler and then retrieves completed TaskRun
+func getNewTaskRun(taskRun v1beta1.TaskRun, d []*v1beta1.TaskRun, t *testing.T) (*v1beta1.TaskRun, error) {
+	testAssets, cancel := getFakeClients(t, d)
```

Comment:
If I'm reading this correctly, this creates new controller/client instances each time this function is called. This feels like it should be done once per setup, and we should simply pass the clients into this func to do the reconcile + Get. What do you think?

Reply:
We talked over meeting that we could use SeedTestData(), but I looked at code, this function() is exactly the one creating client and overwriting old one, plus to use it, we have to create some other configurations. So if we use it, it is basically the same as calling getFakeClients(). I also searched for other code in Tekton repository, for different data in the same test, new client is created for each data set as well.